src/sentry/hybridcloud/rpc/services/caching/service.py

```diff
@@ -91,23 +98,137 @@ def resolve_from(
+        values = _consume_generator(_get_cache([key], self.silo_mode))
+        return _consume_generator(self.resolve_from(object_id, values))
+
+
+class SiloCacheManyBackedCallable(Generic[_R]):
+    """
+    Get a multiple records from cache or wrapped function.
+
+    When cache read returns no or partial data, the wrapped function will be invoked
+    with keys missing data. The result of the wrapped function will then be stored in cache.
+    """
+
+    silo_mode: SiloMode
+    base_key: str
+    cb: Callable[[list[int]], list[_R]]
+    type_: type[_R]
+    timeout: int | None
+
+    def __init__(
+        self,
+        base_key: str,
+        silo_mode: SiloMode,
+        cb: Callable[[list[int]], list[_R]],
+        t: type[_R],
+        timeout: int | None = None,
+    ):
+        self.base_key = base_key
+        self.silo_mode = silo_mode
+        self.cb = cb
+        self.type_ = t
+        self.timeout = timeout
+
+    def __call__(self, ids: list[int]) -> list[_R]:
+        if (
+            SiloMode.get_current_mode() != self.silo_mode
+            and SiloMode.get_current_mode() != SiloMode.MONOLITH
+        ):
+            return self.cb(ids)
+        return self.get_many(ids)
+
+    def key_from(self, object_id: int) -> str:
+        return f"{self.base_key}:{object_id}"
+
+    def get_many(self, ids: list[int]) -> list[_R]:
+        from .impl import _consume_generator, _delete_cache, _get_cache, _set_cache
+
+        keys = {i: self.key_from(i) for i in ids}
+        cache_values = _consume_generator(_get_cache(list(keys.values()), self.silo_mode))
+
+        # Mapping between object_id and cache versions
+        missing: dict[int, int] = {}
+        found: dict[int, _R] = {}
+
+        for object_id, cache_key in keys.items():
+            version: int | None = None
+            cache_value = cache_values[cache_key]
+            if isinstance(cache_value, str):
+                # Found data in cache
+                try:
+                    found[object_id] = self.type_(**json.loads(cache_value))
+                except (pydantic.ValidationError, JSONDecodeError, TypeError):
+                    version = _consume_generator(_delete_cache(cache_key, self.silo_mode))
+            else:
+                # Data was missing in cache but we have a version for the cache key
+                version = cache_value
+            if version is not None:
+                missing[object_id] = version
+
+        # This result could have different order than missing_object_ids, or have gaps
+        cb_result = self.cb(list(missing.keys()))
+        for record in cb_result:
+            # TODO(hybridcloud) The types/interfaces don't make reading this attribute safe.
+            # We rely on a convention of records having `id` for now. In the future
+            # this could become a decorator parameter intead.
+            record_id = getattr(record, "id")
+            if record_id is None:
+                continue
```

Comment:
Couldn't this behavior vary per RPC service? I'd imagine some services care about all queried records existing, while others do not.

Reply:
Yeah that could happen. We could set `found[record_id] = None` and let the caller figure out what is going on. 🤷 I chose to go with the skip behavior, as the behavior of `user_service.get_many()` is to return nothing for ids that don't exist. We have similar behavior in other get_many methods.